monitoring/etcd/etcd.go

```diff
@@ -0,0 +1,189 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gravitational/satellite/monitoring"
+	"github.com/gravitational/trace"
+	"github.com/prometheus/client_golang/prometheus"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const (
+	namespace             = "etcd"
+	collectMetricsTimeout = 5 * time.Second
+)
+
+var (
+	followersLatency = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "etcd_followers_latency"),
+		"Bucketed histogram of latency time (s) between ETCD leader and follower",
+		//			Buckets:   prometheus.ExponentialBuckets(0.0005, 2, 13), // buckets from 0.0001 till 4.096 sec
+		[]string{"followerName"}, nil,
+	)
+
+	followersRaftFail = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "etcd_followers_raft_fail"),
+		"Counter of Raft RPC failed requests between ETCD leader and follower",
+		[]string{"followerName"}, nil,
+	)
+
+	followersRaftSuccess = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "etcd_followers_raft_success"),
+		"Counter of Raft RPC successful requests between ETCD leader and follower",
+		[]string{"followerName"}, nil,
+	)
+)
+
+// LeaderStats is used by the leader in an etcd cluster, and encapsulates
+// statistics about communication with its followers
+// reference documentation https://github.com/coreos/etcd/blob/master/etcdserver/stats/leader.go
+type LeaderStats struct {
+	// Leader is the ID of the leader in the etcd cluster.
+	Leader    string                    `json:"leader"`
+	Followers map[string]*FollowerStats `json:"followers"`
+	Message   string                    `json:"message"`
+}
+
+// FollowerStats encapsulates various statistics about a follower in an etcd cluster
+type FollowerStats struct {
+	Latency   LatencyStats `json:"latency"`
+	RaftStats RaftStats    `json:"counts"`
+}
+
+// LatencyStats encapsulates latency statistics.
+type LatencyStats struct {
+	Current float64 `json:"current"`
+}
+
+// RaftStats encapsulates raft statistics.
+type RaftStats struct {
+	Fail    uint64 `json:"fail"`
+	Success uint64 `json:"success"`
+}
+
+// Exporter collects ETCD stats from the given server and exports them using
+// the prometheus metrics package.
+type Exporter struct {
+	client *http.Client
+	config monitoring.ETCDConfig
+	mutex  sync.RWMutex
+
+	followersLatency     *prometheus.GaugeVec
+	followersRaftFail    *prometheus.GaugeVec
+	followersRaftSuccess *prometheus.GaugeVec
+}
+
+// NewExporter returns an initialized ETCDExporter.
+func NewExporter(config *monitoring.ETCDConfig) (*Exporter, error) {
+	transport, err := config.NewHTTPTransport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	client := &http.Client{
+		Transport: transport,
+		Timeout:   collectMetricsTimeout,
+	}
```

Comment:
What I meant was that you could:
```go
// TODO: ensure there's at least a single etcd endpoint configured
client, err := roundtrip.NewClient(config.Endpoints[0], "v2")
// ... and later
resp, err := client.Get(client.Endpoint("stats", "leader"))
// use resp.Body() w/o having to close or otherwise manage it
```